### DISCUSSION TITLE
[Problem] Not starting a new process with runner.run() ?

### DISCUSSION THREAD

**Discussion by twardoch** [fonttools/fontbakery#4389] [Q&A]
I can run this code: 

```python
#!/usr/bin/env python3
from fontbakery.checkrunner import CheckRunner, get_module_profile
from fontbakery.commands.check_profile import get_module
from fontbakery.configuration import Configuration
from fontbakery.multiproc import multiprocessing_runner
from fontbakery.reporters.html import HTMLReporter
from fontbakery.status import WARN


def get_fontbakery_html(font_paths, profile_name="universal"):
    profile = get_module_profile(get_module(f"fontbakery.profiles.{profile_name}"))
    config = Configuration(full_lists=True)
    runner = CheckRunner(profile, values={"fonts": font_paths}, config=config)
    hr = HTMLReporter(runner=runner, loglevels=[WARN])
    hr_callback = hr.receive
    gen = runner.run()
    for item in gen:
        hr_callback(item)
    return hr.get_html()


def main(font_path, profile_name="universal"):
    return get_fontbakery_html([font_path], profile_name)


print(main("/Users/adam/Desktop/Nabla/Web-TT/Nabla-Regular-COLRv1.ttf"))
```

in Terminal, and it works just fine. 

But when I run the same code in FontLab 8, FontLab runs another instance of FontLab in the line that has `runner.run()`, and then waits infinitely until I quit the other instance of FontLab. This is because FontLab 8 is the "interpreter" executable for FontLab scripts, and generally this happens if I start a new process from Python. 

This indicates that somewhere in the code, FontBakery is running another instance (process) of the Python interpreter. 

1. Where can this happen? 
2. How do I prevent this?

**Comment by twardoch** [comment]
I understand where this happens if I use `multiprocessing_runner` but I don’t quite understand how a new process gets spun off if I just use `runner.run()`.

**Reply by davelab6** [reply]
@graphicore might have some insight

**Reply by graphicore** [reply]
I have no good idea what's causing this. All the multi-processing done by `multiprocessing_runner` is in the `multiproc` module. It shouldn't do these things if you don't call that code.

But maybe some check creates a sub-process? You could try to run with less checks, or maybe with just one check at a time and see if some checks run and some checks block.
For debugging, can you identify the child processes created by FontLab? How do they get called? When I run e.g. `$ fontbakery check-googlefonts -j ./fonts/ofl/rubik/*` and monitor it using the `htop` command (with f4 FILTER: fontbakery), I see all the sub processes called with the same command, but sometimes a call to `$ /home/commander/Projects/googlefonts/fontbakery/venv/bin/python3 -m pip list -v --format=json` pops up. I think that might come from `com.google.fonts/check/fontbakery_version` in `profiles/universal.py`, Same subprocess also without multiprocessing/`-j`.

![Screenshot from 2023-12-19 11-26-36](https://github.com/fonttools/fontbakery/assets/393132/fd2190f4-acb3-49f1-817c-752588f0a5ba)

**Reply by graphicore** [reply]
If you can use `TerminalReporter` set `print_progress=false` and maybe set `stdout` to a  python file-like that writes to a file, you should get a progress report until the blocking position, including the last output for `STARTCHECK` which happens before the  check itself is executed.

**Reply by graphicore** [reply]
cc @twardoch @davelab6: direct mentioning you as I don't know if the discussions feature would inform you when I post in a sub-thread. Also, should this be an issue?
